### ISSUE TITLE
Add support for non-Message events

### ISSUE THREAD

**Issue by Cadair** [opsdroid/opsdroid#536]
Currently the main opsdroid bus only handles connectors receiving messages and passing them to skills as `Message` objects. For opsdroid to be used to do things other than a chat bot it would be good if there was support for generic event types, which could be defined on a per-connector level.

Examples of these events that have come to mind:

- A `NewChannel` event in the slack connector.
- A `RoomInvite` event in the matrix connector
- An `Issue` or `PullRequest` or `Comment` or `Review` event in the github connector (or a new GitHub App connector).

I think that the important thing with these events will be that the connector can define and emit them, and any skill can match to any event (maybe also matching on properties of event type). The simple way I have thought to do this is import the event types from the connector, but due to the way opsdroid installs it's own modules this seems to be non-trivial.

**Comment by jacobtomlinson** [comment]
I agree that this would be a useful thing to add. 

This is probably also related to #195 which covers sending different types of responses other than text such as images or binaries.

From an OOP point of view I feel like a message is probably a particular type of event and so we should probably create an `Event` class which `Message` inherits from. 

We should also think about the matchers as they are all for processing `Message` objects. We could add a `match_event` matcher which takes strings like `NewChannel`, `RoomInvite`, etc as arguments. We could also need to add a `process_event` method to opsdroid which would be similar to the `parse` method. Connectors would call `process_event` when they get an event and opsdroid would then run through the skills with `match_event` and run the skill function if the event string matches.

**Comment by Cadair** [comment]
> This is probably also related to #195 which covers sending different types of responses other than text such as images or binaries.

This looks like a nice way to solve this problem, `Image` could be a subclass of `Message`?

> From an OOP point of view I feel like a message is probably a particular type of event and so we should probably create an `Event` class which `Message` inherits from.

Agreed, the current `match_always` matcher then becomes `match_event('Message')`? (Where as `match_always` would be for all events? (backwards compatibility not withstanding))

> We could add a match_event matcher which takes strings like NewChannel, RoomInvite, etc as arguments.

I feel like it would be better if instead of strings it could take the actual classes, but I can't think of a nice way to get hold of the classes in the skill?

I agree with `process_event` etc.

**Comment by SolarDrew** [comment]
Perhaps it could take multiple properties as a tuple only run if they all match? `or`s could be done by just specifying multiple matchers, so it wouldn't need to handle both, though I don't know if having different methods for `and` and `or` is particularly intuitive....

**Comment by jacobtomlinson** [comment]
This functionality sounds like it would fit into the idea of `matchers` and `constraints` which are raised in #273.

```
@match_event(PRReview)
@constraint_event_type("Approve")
def handle_approved_pr(...
```

However I can't immediately think of how to handle logical or's. I guess matchers and constraints could be used as a high level first pass filter and if you want to do more complex logic you can do it in the skill itself.

We could definitely use classes in a better way. This also opens up questions about how responses currently work and whether they could be better. For example calling `PRReview.respond(...)` would probably respond with a `Message` object, but it could also be an `Image` or something else. We could do come type inference style thing where if you respond with a string it infers it should construct a `Message` with the same properties as the message being responded to, but if you pass an instance of `Message` to it you are explicitly defining the properties.

**Comment by Cadair** [comment]
ok, I don't want to drift too far off into constraints and such. I think it would be worth getting just the event matching stuff in first.

@jacobtomlinson do you have a suggestion on a way to "import" the event classes into the skills or are we going to have to use strings?

**Comment by jacobtomlinson** [comment]
My preference would be for strings. That way we can make things more generic. For example both the hypothetical GitHub and Bitbucket connectors could have a `pullrequest` event.

**Comment by Cadair** [comment]
I am not completely adverse to strings, however, I have found in some of my more obscure skill development that it's very useful to be connector dependant when you need to be (see [skill-picard](https://github.com/SolarDrew/skill-picard) which is *very* matrix specific). So while I like the idea of being able to match on some kind of `.name` property, I would also like to be able to match on the type of the event as well, i.e. `GitHubPR` vs `BitBucketPR` as subclasses of events emitted by the respective skills.

**Comment by jacobtomlinson** [comment]
Yeah namespacing would be good. Perhaps with dot notation? `github.pr` or `bitbucket.pr`

This is the format I tend to follow in the NLU APIs like Rasa and Dialogflow.
